tools/sense_studio/train_module.py

```diff
@@ -0,0 +1,84 @@
+import os
+import subprocess
+import shlex
+import urllib
+import time
+
+import flask
+from flask import Blueprint
+from flask import render_template
+from flask import request
+from flask import current_app
+from flask import stream_with_context
+
+from tools.sense_studio import utils
+
+train_bp = Blueprint('train_bp', __name__)
+
+PROCESS = None
+
+
+@train_bp.route('/<string:project>', methods=['GET'])
+def training_page(project):
+    project = urllib.parse.unquote(project)
+    return render_template('train.html', project=project, models=utils.BACKBONE_MODELS, is_disabled=False)
+
+
+def stream_template(template_name, **context):
+    # Ref: https://flask.palletsprojects.com/en/1.1.x/patterns/streaming/
+    current_app.update_template_context(context)
+    t = current_app.jinja_env.get_template(template_name)
+    rv = t.stream(context)
+    return rv
+
+
+@train_bp.route('/train-model', methods=['POST'])
+def train_model():
+    data = request.form
+    project = data['project']
+    num_layers_to_finetune = data['layers_to_finetune']
+    path_out = data['output_folder']
+    model_name = data['model_name']
+    path = utils.lookup_project_path(project)
+    config = utils.load_project_config(path)
+
+    train_classifier = ["python tools/train_classifier.py", f"--path_in=/home/twentybn/Code/sense/dataset/{project}",
+                        f"--num_layers_to_finetune={num_layers_to_finetune}",
+                        "--use_gpu" if config['use_gpu'] else "",
+                        f"--path_out={path_out}" if path_out else "",
+                        f"--model_name={model_name}" if model_name else "",
+                        "--overwrite"]
+
+    train_classifier = ' '.join(train_classifier)
+    train_classifier = shlex.split(train_classifier)
+
+    global PROCESS
+    PROCESS = subprocess.Popen(train_classifier, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=False)
+
+    def generate():
+        while True:
+            output = PROCESS.stdout.readline()
+            if output == b'' and PROCESS.poll() is not None:
```

Comment:
Can you explain that end condition?

Reply:
Here, the readline() will give you an encoded string in return, and `PROCESS.poll()` returns `None` if the process is still running, if the process is completed then it will return anything like `-1` or `0`. 
So here this `and` between them helps if there is an empty line in the console log but the processing is not complete yet it will still going on.

Reply:
So wouldn't it be enough to only check `PROCESS.poll() is not None`, or is there a case where `poll` returns `None` but `readline` returned a non-empty string?

Reply:
Yes, there is a case. It will happen when processing is done but the console logs are being saved in `stdout` buffer from which we are reading line one by one for display. So, here processing and reading from buffer are independent processes, so it can happen that process might be complete but we still have some lines left to read from stdout.